test/integration/specs/kbaseCellToolbarMenu_test.js

```diff
@@ -0,0 +1,327 @@
+        },
+        TEST_CASE4: {
+            narrativeId: 58675,
+        },
+        TEST_CASE5: {
+            narrativeId: 58675,
+        }
+    },
+    'narrative-dev': {
+        TEST_CASE1: {
+            narrativeId: 80970,
+        },
+        TEST_CASE2: {
+            narrativeId: 80970,
+        },
+        TEST_CASE3: {
+            narrativeId: 80970,
+        },
+        TEST_CASE4: {
+            narrativeId: 80970,
+        },
+        TEST_CASE5: {
+            narrativeId: 80970,
+        }
+    }
+};
+
+const testCases = mergeObjects([testData[browser.config.testParams.ENV], testData.all]);
+
+function mergeObjects(listOfObjects) {
+    const simpleObjectPrototype = Object.getPrototypeOf({});
+
+    function isSimpleObject(obj) {
+        return Object.getPrototypeOf(obj) === simpleObjectPrototype;
+    }
+
+    function merge(obj1, obj2, keyStack) {
+        Object.keys(obj2).forEach(function (key) {
+            const obj1Value = obj1[key];
+            const obj2Value = obj2[key];
+            const obj1Type = typeof obj1Value;
+            // var obj2Type = typeof obj2Value;
+            if (obj1Type === 'undefined') {
+                obj1[key] = obj2[key];
+            } else if (isSimpleObject(obj1Value) && isSimpleObject(obj2Value)) {
+                keyStack.push(key);
+                merge(obj1Value, obj2Value, keyStack);
+                keyStack.pop();
+            } else {
+                console.error('UNMERGABLE', obj1Type, obj1Value);
+                throw new Error('Unmergable at ' + keyStack.join('.') + ':' + key);
+            }
+        });
+    }
+
+    const base = JSON.parse(JSON.stringify(listOfObjects[0]));
+    for (let i = 1; i < listOfObjects.length; i += 1) {
+        merge(base, listOfObjects[i], []);
+    }
+    return base;
+}
+
+async function waitForCell(notebookContainer, cellIndex){
+    return await browser.waitUntil(async () => {
+        const cell = await notebookContainer.$(`.cell:nth-child(${cellIndex})`);
+        return cell;
+    });
+}
+
+async function waitForCellWithTitle(container, cellIndex, title) {
+    const cell = await waitForCell(container, cellIndex);
+    await browser.waitUntil(async () => {
+        const titleElement = await cell.$('[data-element="title"]');
+        const text = await titleElement.getText();
+        return text === title;
+    });
+    return cell;
+}
+
+async function waitForCellWithBody(container, cellIndex, bodyText) {
+    const cell = await waitForCell(container, cellIndex);
+    await browser.waitUntil(async () => {
+        const element = await cell.$('.text_cell_render.rendered_html');
+        const text = await element.getText();
+        return text === bodyText;
+    });
+    return cell;
+}
+
+async function selectCell(container, cellIndex, title) {
+    const cell = await waitForCellWithTitle(container, cellIndex, title);
+
+    // Make sure not selected.
+    // We do this by inspecting the border.
```

Comment:
This seems like a rather brittle way to test the selection status of the cell; would it be better to check what classes the cell has instead? That way at least the tests aren't sensitive to small UI changes.

Reply:
In this case, the test is actually to ensure that the cell being targeted also has the correct content, in this case title. It is first selected by position (index) and then the text of the title area is checked. This relieves the burden on test from having to know the precise selector for locating a cell title (or body, the other variant of this method.)

This function has moved in the latest set of changes, but there is also a method to wait for a cell purely by the index.

Reply:
I've moved this comment to the new location of this code and clarified what I meant.